lib/linter/linter.js

```diff
@@ -542,12 +547,10 @@ function resolveParserOptions(parserName, providedOptions, enabledEnvironments)
         mergedParserOptions.ecmaFeatures = Object.assign({}, mergedParserOptions.ecmaFeatures, { globalReturn: false });
     }
 
-    /*
-     * TODO: @aladdin-add
-     * 1. for a 3rd-party parser, do not normalize parserOptions
-     * 2. for espree, no need to do this (espree will do it)
-     */
-    mergedParserOptions.ecmaVersion = normalizeEcmaVersion(mergedParserOptions.ecmaVersion);
+    // for a 3rd-party parser, do not normalize parserOptions
+    if (parserName === DEFAULT_PARSER_NAME || parserName === void 0) {
```

Comment:
This could be a problem for rules that expect normalized `ecmaVersion`, like `comma-dangle`:

https://github.com/eslint/eslint/blob/85a2725b1fade5538e727102d9701ccb503e54d4/lib/rules/comma-dangle.js#L53

Reply:
Agree. We should always normalize ecmaVersion.

Reply:
I have not thought it was used in rules, will update later.

Reply:
it could be another problem to normalize a 3rd-party parser `ecmaVersion`. as @mdjermanovic 
 mentioned [here](https://github.com/eslint/eslint/pull/14622#discussion_r640580610) .

the problem is eslint cannot tell a 3rd-party parser's default, and "latest".

Reply:
after some thinking, I would suggest not to change it.

If it's an issue, it has existed in the previous releases. The change is not a bugfix, let's make another change to fix it.

Reply:
Can we retain the exact same behavior as it is now for third-party parsers, and make changes only if it's default parser?

Reply:
pushed a commit 99604d4e1617bda1622426c8efd30aa397de8c81 to revert "always normalize ecmaVersion".

Reply:
In my opinion we should always do the following, regardless of the parser, because that's the current behavior:

https://github.com/eslint/eslint/blob/ec28b5a2bdc69f34ce29d670f5e84d2446774a00/lib/linter/linter.js#L444

But the new normalizations (`undefined` -> `5`, `"latest"` -> `espree.latestEcmaVersion`) should be done only when `espree` is used.

Reply:
👍 good catch! I missed it somehow. will update

Reply:
should be addressed now. thanks!